### PR TITLE
Refactoring & adding tests to rescorer in the example plugin, to account for non strictly increasing doc ids in the results.

### LINKED ISSUE #39665: Example rescore plugin sorted order is not used
Looking at the comment on the example rescore plugin here:

https://github.com/elastic/elasticsearch/blob/93034aff49df8ae0b6b15c311b0f369357f86534/plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java#L163-L175

The `sortedByDocId` variable is not used anywhere. Is this comment vestigial, or the pipes not connected?

While on this topic, can this loop be explained?

https://github.com/elastic/elasticsearch/blob/93034aff49df8ae0b6b15c311b0f369357f86534/plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java#L181-L185

### PR DESCRIPTION
**Pull request by pmpailis**
In this PR we address an issue with the rescorer example provided where we do not take into account the sorted order of returned docs, when operating on a custom `factor_field` (i.e. field whose value is used for boosting).  This causes some docs to be overlooked if for example `ranking(d_i) < ranking(d_j)` and `d_i.doc > d_j.doc`, i.e. if a document ranked lower in the returned ranked list, has also a smaller docid that is present in a different, already visited, leaf. This results in the doc to be overlooked and the reranker to fail to account for all the results in the provided `window_size`. 

We also add a couple of tests to account for that behavior, by iteratively indexing & refreshing the index, and then perform a search whose top results are in decreasing doc_id order.  

Closes https://github.com/elastic/elasticsearch/issues/39665

### CODE REVIEW

File: plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java [hunk 1]
```diff
@@ -163,34 +163,36 @@ public TopDocs rescore(TopDocs topDocs, IndexSearcher searcher, RescoreContext r
...
+                    ScoreDoc scoreDoc = sortedByDocId[i];
+                    if (scoreDoc.doc >= endDoc) {
                         do {
                             leaf = leaves.next();
                             endDoc = leaf.docBase + leaf.reader().maxDoc();
-                        } while (topDocs.scoreDocs[i].doc >= endDoc);
+                        } while (scoreDoc.doc >= endDoc);
                         LeafFieldData fd = context.factorField.load(leaf);
-                        if (false == (fd instanceof LeafNumericFieldData)) {
+                        if (!(fd instanceof LeafNumericFieldData)) {
```
Comment (benwtrent) [inline plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java:181]: `== false` is preferred. Though I think the example plugins don't fall under our formatting rules.

File: plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java [hunk 2]
```diff
@@ -163,34 +163,36 @@ public TopDocs rescore(TopDocs topDocs, IndexSearcher searcher, RescoreContext r
...
                             throw new IllegalArgumentException("[" + context.factorField.getFieldName() + "] is not a number");
                         }
                         data = ((LeafNumericFieldData) fd).getDoubleValues();
                     }
-                    if (false == data.advanceExact(topDocs.scoreDocs[i].doc - leaf.docBase)) {
-                        throw new IllegalArgumentException("document [" + topDocs.scoreDocs[i].doc
-                                + "] does not have the field [" + context.factorField.getFieldName() + "]");
+                    assert data != null;
+                    if (!data.advanceExact(scoreDoc.doc - leaf.docBase)) {
```
Comment (benwtrent) [inline plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java:187]: `== false` is preferred. Though I think the example plugins don't fall under our formatting rules.

**Review by benwtrent — APPROVED** [review]

File: plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java (hunk 1 above)
Comment (pmpailis) [reply plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java]: Reverted back to `== false` format.

File: plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java (hunk 2 above)
Comment (pmpailis) [reply plugins/examples/rescore/src/main/java/org/elasticsearch/example/rescore/ExampleRescoreBuilder.java]: Reverted back to == false format.

**Pull request merged**